ruby/sdk-phase-1.md

```diff
@@ -0,0 +1,439 @@
+    cross-language compatible.
+  * There are not any good JSON-to-object conversion approaches in Ruby beyond this.
+  * Therefore, very unfortunately, it is unreasonable at this time for any area where conversion from payload to value
+    is needed to specify the type to convert to. Users can use type hints in their RBS files and such, but that has no
+    runtime check/effect.
+* The JSON encoding converter initializer will have a way to set parse and generate options.
+* The payload converter will have a way to let the default be created with JSON parse and generate options for the JSON
+  encoding converter.
+  * 💭 Why? Users wanting to simply set JSON options shouldn't have to go through the mess of recreating the default
+    list of encoding converters.
+
+## Activities
+
+A simple, heartbeating activity that calls a separate thing:
+
+```ruby
+class MyActivity < Temporalio::Activity
+  # Can customize name, e.g.
+  # activity_name 'custom-name'
+
+  # Can customize executor, e.g.
+  # activity_executor :fiber
+
+  # Can make an activity dynamic, e.g.
+  # activity_dynamic true
+
+  def initialize(db_client)
+    @db_client = db_client
+  end
+
+  def execute(arg)
+    loop do
+      db_result = @db_client.do_thing(arg)
+      if db_result
+        return db_result
+      end
+      sleep 1
+      Temporalio::Activity::Context.heartbeat
+    end
+  end
+end
+```
+
+* Activities are defined as classes.
+  * 💭 Why classes instead of methods/functions like every other SDK? There are tradeoffs to both approaches. As
+    classes, they are more easily referenced, a bit more easily typed in RBS, and metadata can be applied more easily at
+    the class level. This comes at a cost of not being able to easily share state across classes, but users can provide
+    something in the constructor multiple activities use. Open to considering activity methods instead.
+  * ❓ Would we prefer activities as methods knowing they may not be able to be typed/referenced well? We can't really
+    take the workflow signal-ish approach of those pseudo-decorators making class methods unless we say activities must
+    be instance methods (since we would allow class methods as activities).
+* Activity classes are instantiated at registration time, not for each invocation.
+  * 💭 Why? Existing Temporal Ruby SDK does instantiate for each activity task, but this discourages use of shared
+    objects or state to be referenced. This does have the tradeoff our other SDKs have of requiring the function to be
+    thread safe.
+* `Temporalio::Activity::Context` is a thread/fiber-local context available to activities.
+  * It has instance methods for info, heartbeat, etc.
+  * It has a class method for `current` that returns from thread/fiber-local and class method shortcuts for all instance
+    methods.
+    * 💭 Why class methods equivalents? `Temporalio::Activity::Context.heartbeat` is easier to read than
```

Comment:
In other SDKs, `heartbeat`, `log`, `info`, etc... is on the `activity` module.
Why not do that here too? E.g:

```
Temporalio::Activity::Context.current

Temporalio::Activity::heartbeat
Temporalio::Activity::info
Temporalio::Activity::log
```

Reply:
Because `Temporalio::Activity` is the class that is being extended by the activity implementation and we don't want people to think that this is only usable in that extension nor do we want to clutter the inherited class with methods like these when they are part of the context.